### PR TITLE
fix: Workaround Ruby bug in 2.7.3 with kwrest

### PR DESCRIPTION
**Pull request by kgilpin**
There are two main elements to this fix:

* Call https://ruby-doc.org/core-2.7.0/Proc.html#method-i-ruby2_keywords
* Fix a probable side-effect of using ruby_keywords: https://github.com/applandinc/appmap-ruby/pull/154/files#diff-95a15f6e0d63580f2d31bf5a2e866f1062fbda51da73ad9089c579b952467a08R54

Fixes https://github.com/applandinc/appmap-ruby/issues/153

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,5 +1,10 @@
 language: ruby
 
+rbenv:
```
Comment (kgilpin) [inline .travis.yml:3]: We go to a build matrix scheme in order to run the non-Dockerized tests across Ruby versions.

File: Rakefile [hunk 2]
```diff
@@ -26,15 +25,25 @@ namespace 'gem' do
...
+        sh('yarn install --prod')
   
         default_build_gem
       end
     end
   end
 end
 
-RUBY_VERSIONS=%w[2.5 2.6 2.7]
+RUBY_VERSIONS=%w[2.5 2.6 2.7].select do |version|
```
Comment (kgilpin) [inline Rakefile:36]: This is the simplest way to get the behavior we want without doing a major build refactor.

File: appmap.gemspec [hunk 3]
```diff
@@ -54,4 +54,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'webdrivers', '~> 4.0'
   spec.add_development_dependency 'timecop'
   spec.add_development_dependency 'hashie'
+  spec.add_development_dependency 'webrick'
```
Comment (kgilpin) [inline appmap.gemspec:57]: This is a separate Gem in Ruby 3. Preparing for that.

File: lib/appmap/handler/net_http.rb [hunk 4]
```diff
@@ -82,7 +82,7 @@ class << self
         def request_headers(request)
           {}.tap do |headers|
             request.each_header do |k,v|
-              key = [ 'HTTP', k.underscore.upcase ].join('_')
+              key = [ 'HTTP', Util.underscore(k).upcase ].join('_')
```
Comment (kgilpin) [inline lib/appmap/handler/net_http.rb:85]: Moving away from activesupport.

File: lib/appmap/handler/rails/template.rb [hunk 5]
```diff
@@ -146,7 +146,8 @@ def handle_return(call_event_id, elapsed, return_value, exception)
         class RenderHandler
           class << self
             def handle_call(defined_class, hook_method, receiver, args)
-              context, options = args
+              # context, options
```
Comment (kgilpin) [inline lib/appmap/handler/rails/template.rb:149]: Fix an unused argument warning.

File: lib/appmap/hook.rb [hunk 6]
```diff
@@ -99,7 +99,6 @@ def trace_location(trace_point)
     end
 
     def trace_end(trace_point)
-      location = trace_location(trace_point)
```
Comment (kgilpin) [inline lib/appmap/hook.rb:102]: Fix an unused argument warning.

File: lib/appmap/hook/method.rb [hunk 7]
```diff
@@ -45,42 +47,48 @@ def activate
...
-        hook_method_def = nil
-        hook_class.instance_eval do 
-          hook_method_def = Proc.new do |*args, &block|
-            instance_method = hook_method.bind(self).to_proc
-            call_instance_method = -> { instance_method.call(*args, &block) }
+        hook_method_def = Proc.new do |*args, &block|
+          instance_method = hook_method.bind(self).to_proc
+          call_instance_method = -> {
+            # https://github.com/applandinc/appmap-ruby/issues/153
+            if Util.ruby_minor_version >= 2.7 && args == ARRAY_OF_EMPTY_HASH && hook_method.arity == 1
```
Comment (kgilpin) [inline lib/appmap/hook/method.rb:54]: This is the actual fix for the bug. In this particular case, `instance_method.call(*args, &block)` will actually pass 0 args to the instance_method. We detect that case and make the correct call.

File: lib/appmap/hook/method.rb [hunk 8]
```diff
@@ -45,42 +47,48 @@ def activate
...
+          rescue
+            exception = $ERROR_INFO
+            raise
+          ensure
+            with_disabled_hook.call do
+              after_hook.call(self, call_event, start_time, return_value, exception) if call_event
             end
           end
         end
+        hook_method_def = hook_method_def.ruby2_keywords if Util.ruby_minor_version >= 2.7
```
Comment (kgilpin) [inline lib/appmap/hook/method.rb:90]: This is probably how the problem above is introduced - but it's necessary to have the proper behavior in Ruby 2.7.

File: lib/appmap/trace.rb [hunk 9]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'delegate'
```
Comment (kgilpin) [inline lib/appmap/trace.rb:3]: Turns out this is supposed to be a separate import.

File: lib/appmap/util.rb [hunk 10]
```diff
@@ -124,7 +124,7 @@ def swaggerize_path(path)
         path = path.split('(.')[0]
         tokens = path.split('/')
         tokens.map do |token|
-          token.gsub /^:(.*)/, '{\1}'
+          token.gsub(/^:(.*)/, '{\1}')
```
Comment (kgilpin) [inline lib/appmap/util.rb:127]: This was a warning.

File: spec/record_net_http_spec.rb [hunk 11]
```diff
@@ -1,6 +1,7 @@
 require 'spec_helper'
 require 'diffy'
 require 'rack'
+require 'webrick'
```
Comment (kgilpin) [inline spec/record_net_http_spec.rb:4]: Webrick is a separate gem in Ruby 3.0.

File: lib/appmap/hook/method.rb (hunk 8 above)
Comment (apotterri) [reply lib/appmap/hook/method.rb:90]: The doc suggests using `respond_to?(:ruby2_keywords)`. Is there any reason to prefer that?

Comment (apotterri) [reply lib/appmap/hook/method.rb:90]: So, you're saying this causes a problem, and the change above works around the problem, right? Can you give an example of the bad behavior this change fixes, please?

**Review by apotterri — COMMENTED** [review]
I'd like understand the results of your investigation and how the 2.7-specific changes work a bit better. Other changes here look good.

File: lib/appmap/hook/method.rb (hunk 8 above)
Comment (kgilpin) [reply lib/appmap/hook/method.rb:90]: So, without calling `ruby2_keywords`, then in the case above, args is `[]` and not `[{}]`, so there's really no way to know what the argument to the proxied method should be. At least with `ruby2_keywords`, it's possible to determine what should be happening.

Comment (kgilpin) [reply lib/appmap/hook/method.rb]: We can use `respond_to?`

**Review by torchello — APPROVED** [review]
Looks good! Suggestion: should we try keeping PRs smaller for easier review? (separate fixes in separate PRs)

**Comment by kgilpin** [comment]
Large PRs are a problem, but to me this isn’t that big.

**Review by apotterri — APPROVED** [review]

**Pull request merged**